lib/Dialect/IR/TcpTilingInterfaceImpl.cpp

```diff
@@ -0,0 +1,164 @@
+    OpFoldResult zero = b.getIndexAttr(0);
+    OpFoldResult one = b.getIndexAttr(1);
+    SmallVector<Range> loopRanges(sliceOp.getType().getRank(),
+                                  {zero, one, one});
+    for (auto [idx, size] : llvm::enumerate(sliceOp.getSizes())) {
+      loopRanges[idx].offset = zero;
+      loopRanges[idx].size = size;
+    }
+    return loopRanges;
+  }
+
+  FailureOr<TilingResult>
+  getTiledImplementation(Operation *op, OpBuilder &b,
+                         ArrayRef<OpFoldResult> offsets,
+                         ArrayRef<OpFoldResult> sizes) const {
+    // getTiledImplementation and generateResultTileValue are the same
+    // because:
+    // 1. `getIterationDomain` returns a loop with zero offsets
+    //     and the result slice length
+    // 2. `getResultTilePosition` returns the same offsets and sizes as
+    //     the input offsets and slices.
+    return generateResultTileValue(op, b, 0, offsets, sizes);
+  }
+
+  FailureOr<TilingResult>
+  generateResultTileValue(Operation *op, OpBuilder &b, unsigned resultNumber,
+                          ArrayRef<OpFoldResult> offsets,
+                          ArrayRef<OpFoldResult> sizes) const {
+
+    auto sliceOp = cast<tcp::SliceOp>(op);
+    auto loc = op->getLoc();
+
+    auto fold = [&](OpFoldResult ofr) {
+      return getValueOrCreateConstantIndexOp(b, loc, ofr);
+    };
+    auto add = [&](OpFoldResult val1, OpFoldResult val2) {
+      return b.createOrFold<arith::AddIOp>(loc, fold(val1), fold(val2));
+    };
+    auto mul = [&](OpFoldResult val1, OpFoldResult val2) {
+      return b.createOrFold<arith::MulIOp>(loc, fold(val1), fold(val2));
+    };
+
+    // Offset on the input of slice is computed by:
+    //     start + offset_on_output * stride
+    //
+    // Tile and fuse algorithm does not work with stride != 1.
+    // In order to get around this, we extract a contiguous chunk with
+    // `tensor.extract_slice` and use the following `tcp.slice` op to
+    // extract the strided parts. So, the size on the
+    // `tensor.extract_slice` will be
+    //     size * stride
+    auto sliceStart = sliceOp.getStarts();
+    auto sliceStrides = sliceOp.getStrides();
+    SmallVector<OpFoldResult> newOffsets;
+    SmallVector<OpFoldResult> newSizes;
+    SmallVector<OpFoldResult> newStrides;
+    for (auto [offset, start, stride, size] :
+         llvm::zip(offsets, sliceStart, sliceStrides, sizes)) {
+      newOffsets.push_back(add(start, mul(offset, stride)));
+      newSizes.push_back(mul(size, stride));
+      newStrides.push_back(b.createOrFold<arith::ConstantIndexOp>(loc, 1));
+    }
+
+    auto extractOp = b.create<tensor::ExtractSliceOp>(
+        loc, sliceOp.getIn(), newOffsets, newSizes, newStrides);
+
+    // Add a `tcp.slice` op on the tile.
+    auto zero = b.create<arith::ConstantIndexOp>(loc, 0);
+    SmallVector<Value> zeroOffsets(offsets.size(), zero.getResult());
+    auto sizeIntValues = getConstantIntValues(sizes);
+    assert(
```

Comment:
I don't know if this assert would work... What happens if we run `structured.fuse` on back to back `tcp.slice` ops with non-trivial strides for instance? Given that the `tensor.extract_slice` created above has a computed value for the size, this is the size which will be passed to the upstream `tcp.slice`'s tiling interface. Maybe you can give such an example a shot?

If we indeed do not want to support non-constant values for the sizes, this should probably be a condition for which we return failure() at the beginning of the function.

Also, I think all we want to do here is to create a tensor type where the corresponding dim is `kDynamic` for the case when the corresponding size is not a constant integer.

Reply:
Good point. That is not necessary. Changed it to use `kDynamic` when these are not constants.